packages/rescript-relay/src/RescriptRelay.res

```diff
@@ -347,78 +347,65 @@ module ReadOnlyRecordSourceProxy = {
-
-  type rec normalizationListValueArgument = {
-    name: string,
-    items: array<Nullable.t<normalizationArgumentWrapped>>,
-  }
-  and normalizationLiteralArgument = {
-    name: string,
-    @as("type") type_: Nullable.t<string>,
-    value: JSON.t,
-  }
-  and normalizationObjectValueArgument = {
-    name: string,
-    fields: Nullable.t<array<normalizationArgumentWrapped>>,
-  }
-  and normalizationVariableArgument = {
-    name: string,
-    @as("type") type_: Nullable.t<string>,
-    variableName: string,
-  }
-
-  type normalizationArgument =
-    | ListValue(normalizationListValueArgument)
-    | Literal(normalizationLiteralArgument)
-    | ObjectValue(normalizationObjectValueArgument)
-    | Variable(normalizationVariableArgument)
-
-  let unwrapNormalizationArgument = wrapped =>
-    switch wrapped.kind {
-    | #ListValue => ListValue(Obj.magic(wrapped))
-    | #Literal => Literal(Obj.magic(wrapped))
-    | #ObjectValue => ObjectValue(Obj.magic(wrapped))
-    | #Variable => Variable(Obj.magic(wrapped))
-    }
+  @tag("kind")
+  type rec normalizationArgument =
+    | ListValue({name: string, items: array<null<normalizationArgument>>})
+    | Literal({name: string, @as("type") type_: nullable<string>, value: JSON.t})
+    | ObjectValue({name: string, fields: array<normalizationArgument>})
+    | Variable({name: string, @as("type") type_: nullable<string>, variableName: string})
 
   type normalizationScalarField = {
-    alias: Nullable.t<string>,
+    alias: nullable<string>,
     name: string,
-    args: Nullable.t<array<normalizationArgumentWrapped>>,
-    storageKey: Nullable.t<string>,
+    args: nullable<array<normalizationArgument>>,
+    storageKey: nullable<string>,
   }
 
-  let makeScalarMissingFieldHandler = handle =>
-    Obj.magic({
-      "kind": #scalar,
-      "handle": handle,
-    })
-
   type normalizationLinkedField = {
-    alias: Nullable.t<string>,
+    alias: nullable<string>,
     name: string,
-    storageKey: Nullable.t<string>,
-    args: Nullable.t<array<normalizationArgument>>,
-    concreteType: Nullable.t<string>,
+    storageKey: nullable<string>,
+    args: nullable<array<normalizationArgument>>,
+    concreteType: nullable<string>,
     plural: bool,
     selections: array<JSON.t>,
   }
 
-  let makeLinkedMissingFieldHandler = handle =>
-    Obj.magic({
-      "kind": #linked,
-      "handle": handle,
-    })
-
-  let makePluralLinkedMissingFieldHandler = handle =>
-    Obj.magic({
-      "kind": #pluralLinked,
-      "handle": handle,
-    })
+  @tag("kind")
+  type rec t =
+    | @as("scalar")
+    Scalar({
+        handle: (
+          normalizationScalarField,
+          nullable<'record>,
+          'args,
+          ReadOnlyRecordSourceProxy.t,
+        ) => 'scalarValue,
+      }): t
+    | @as("linked")
+    Linked({
+        handle: (
+          normalizationLinkedField,
+          nullable<'record>,
+          'args,
+          ReadOnlyRecordSourceProxy.t,
+        ) => option<dataId>,
```

Comment:
I don't think this is going to work. Unless something has changed, returning Nullable is required here, or you won't be able to distinguish between `null` (fetched but no data) and `undefined` (data not fetched). Same for the plural linked field IIRC.